src/server/game/AI/CoreAI/UnitAI.h

```diff
@@ -152,94 +152,174 @@ class TC_GAME_API UnitAI
         {
-            ThreatContainer::StorageType const& threatlist = GetThreatManager().getThreatList();
-            if (threatlist.empty())
+            targetList.clear();
+            ThreatManager& mgr = GetThreatManager();
+            // shortcut: we're gonna ignore the first <offset> elements, and there's at most <offset> elements, so we ignore them all - nothing to do here
+            if (mgr.GetThreatListSize() <= offset)
                 return;
 
-            for (HostileReference* hostileRef : threatlist)
-                if (predicate(hostileRef->getTarget()))
-                    targetList.push_back(hostileRef->getTarget());
+            if (targetType == SELECT_TARGET_MAXDISTANCE || targetType == SELECT_TARGET_MINDISTANCE)
+            {
+                for (ThreatReference* ref : mgr.GetUnsortedThreatList())
+                {
+                    if (ref->IsOffline())
+                        continue;
+                    targetList.push_back(ref->GetVictim());
+                }
+            }
+            else
+            {
+                Unit* currentVictim = mgr.GetCurrentVictim();
+                if (currentVictim)
+                    targetList.push_back(currentVictim);
+                for (ThreatReference* ref : mgr.GetSortedThreatList())
+                {
+                    if (ref->IsOffline())
+                        continue;
+                    Unit* thisTarget = ref->GetVictim();
+                    if (thisTarget != currentVictim)
+                        targetList.push_back(thisTarget);
+                }
+            }
 
-            if (targetList.size() < maxTargets)
+            // shortcut: the list isn't gonna get any larger
+            if (targetList.size() <= offset)
+            {
+                targetList.clear();
                 return;
+            }
 
-            if (targetType == SELECT_TARGET_NEAREST || targetType == SELECT_TARGET_FARTHEST)
-                SortByDistanceTo(me, targetList);
-
-            if (targetType == SELECT_TARGET_FARTHEST || targetType == SELECT_TARGET_BOTTOMAGGRO)
+            // right now, list is unsorted for DISTANCE types - re-sort by MAXDISTANCE
+            if (targetType == SELECT_TARGET_MAXDISTANCE || targetType == SELECT_TARGET_MINDISTANCE)
+                targetList.sort(Trinity::ObjectDistanceOrderPred(me, false));
+            // now the list is MAX sorted, reverse for MIN types
+            if (targetType == SELECT_TARGET_MINTHREAT || targetType == SELECT_TARGET_MINDISTANCE)
                 targetList.reverse();
+            // ignore the first <offset> elements
+            while (offset)
+            {
+                targetList.pop_front();
+                --offset;
+            }
+            // and filter by predicate
+            for (auto it = targetList.begin(), end = targetList.end(); it != end;)
+            {
+                if (predicate(*it))
```

Comment:
Honestly, if your goal is to get this merged as fast as possible then you should revert this order change (reviewing it would require checking every call to SelectTarget in scripts, something I don't want to do in the scope of this PR)